### PR TITLE
Refactoring of interpolation functions for blinds Prof and Slat Angle

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
This branch includes the refactoring of the listed functions (e.g. inlining, cache intermediate results, reduce redundant calcs).
- InterpSw
- InterBlinds (not used)
- InterpProfAng
- InterpSlatAng
- InterpProfSlatAng
- BlindBeamBeamTrans

Also, the MACROs in checking shading flag status were converted to `constexpr`.

The small diffs with the 3 blind related testfiles result from the floating-point error explained [here](https://github.com/NREL/EnergyPlus/blob/ade8f89debeebfbb1bea32f4b253c5f81d497f93/src/EnergyPlus/SolarShading.cc#L9329). Previously `InterpSlatAng` uses `* DeltaAng_inv;`, while `InterpProfSlatAng` uses `/ DeltaAng` in `state.dataSurface->SurfWinSlatsAngInterpFac(ISurf)` calculation. The diffs are less than 10e-15.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] ~If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix~
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

File: src/EnergyPlus/General.hh [hunk 1]
```diff
@@ -157,6 +157,27 @@ namespace General {
...
     );
 
+    constexpr Real64 InterpGeneral(Real64 const Lower,
+                                Real64 const Upper,
+                                Real64 const InterpFac
+    )
+    {
+        return Lower + InterpFac * (Upper - Lower);
+    };
+
```
Comment (amirroth) [inline src/EnergyPlus/General.hh:143]: I'm sure declaring this function `constexpr` vs. just `inline` doesn't hurt anything, but I doubt any of the arguments will actually be `constexpr` and enable any compile-time reduction. Same for `InterpProfSlat`.

**Comment by xuanluo113** [comment]
@mjwitte This branch is ready for review. Please let me know if you have any availability in reviewing this, or you have some recommendations for reviewers. I also wonder if this could come before the 9.5 release. Thanks!

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc [hunk 2]
```diff
@@ -261,11 +260,8 @@ namespace HeatBalanceIntRadExchange {
                         int const SurfNum = zone_SurfacePtr(ZoneSurfNum);
                         int const ConstrNum = state.dataSurface->Surface(SurfNum).Construction;
                         zone_info.Emissivity(ZoneSurfNum) = state.dataConstruction->Construct(ConstrNum).InsideAbsorpThermal;
-                        auto const &surface_window(state.dataSurface->SurfaceWindow(SurfNum));
                         if (state.dataConstruction->Construct(ConstrNum).TypeIsWindow && ANY_INTERIOR_SHADE_BLIND(state.dataSurface->SurfWinShadingFlag(SurfNum))) {
-                            zone_info.Emissivity(ZoneSurfNum) =
-                                InterpSlatAng(state.dataSurface->SurfWinSlatAngThisTS(SurfNum), state.dataSurface->SurfWinMovableSlats(SurfNum), surface_window.EffShBlindEmiss) +
-                                InterpSlatAng(state.dataSurface->SurfWinSlatAngThisTS(SurfNum), state.dataSurface->SurfWinMovableSlats(SurfNum), surface_window.EffGlassEmiss);
+                            zone_info.Emissivity(ZoneSurfNum) = state.dataHeatBal->ITABSF(SurfNum);
```
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc:264]: So, if I'm following correctly, `ITABSF` is set in `HeatBalanceSurfaceManager::ComputeIntThermalAbsorpFactors` (same as before, with some adjustments in this branch) so no need to recompute it here.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc [hunk 3]
```diff
@@ -318,8 +314,7 @@ namespace HeatBalanceIntRadExchange {
                         // and emiss is used here that is a weighted combination of shade/blind and glass temp and emiss.
                     } else if (ANY_INTERIOR_SHADE_BLIND(state.dataSurface->SurfWinShadingFlag(SurfNum))) {
                         SurfaceTempRad[ZoneSurfNum] = state.dataSurface->SurfWinEffInsSurfTemp(SurfNum);
-                        SurfaceEmiss[ZoneSurfNum] = InterpSlatAng(state.dataSurface->SurfWinSlatAngThisTS(SurfNum), state.dataSurface->SurfWinMovableSlats(SurfNum), surface_window.EffShBlindEmiss) +
-                            InterpSlatAng(state.dataSurface->SurfWinSlatAngThisTS(SurfNum), state.dataSurface->SurfWinMovableSlats(SurfNum), surface_window.EffGlassEmiss);
+                        SurfaceEmiss[ZoneSurfNum] = state.dataHeatBal->ITABSF(SurfNum);
```
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc:317]: Same here, just use `ITABSF` directly. Ok.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 4]
```diff
@@ -2830,35 +2824,53 @@ namespace EnergyPlus::HeatBalanceSurfaceManager {
...
+                                                AbsDiffWinGnd(Lay) = state.dataConstruction->Construct(ConstrNumSh).BlAbsDiffGnd(1, Lay);
+                                                AbsDiffWinSky(Lay) = state.dataConstruction->Construct(ConstrNumSh).BlAbsDiffSky(1, Lay);
+                                            }
                                         }
-                                        state.dataSurface->SurfWinExtDiffAbsByShade(SurfNum) =
-                                                InterpSlatAng(state.dataSurface->SurfWinSlatAngThisTS(SurfNum), state.dataSurface->SurfWinMovableSlats(SurfNum),
-                                                state.dataConstruction->Construct( ConstrNumSh).AbsDiffBlind) *
-                                                (SkySolarInc + GndSolarInc);
+                                        Real64 AbsDiffBlind;
+                                        if (state.dataSurface->SurfWinMovableSlats(SurfNum)) {
```
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Minor detail - these `if` blocks could be combined with the corresponding ones above.

File: src/EnergyPlus/SolarShading.cc [hunk 5]
```diff
@@ -10762,10 +10988,7 @@ namespace EnergyPlus::SolarShading {
 
                         // Accumulate Window and Zone total distributed diffuse solar to check for conservation of energy
                         // For opaque surfaces all incident diffuse is either absorbed or reflected
-//                        WinDifSolarDistAbsorbedTotl += DifSolarAbsW;    // debug [W]
```
Comment (mjwitte) [inline src/EnergyPlus/SolarShading.cc:10765]: Good to clean these out.

File: src/EnergyPlus/DataSurfaces.hh [hunk 6]
```diff
@@ -1513,6 +1538,12 @@ struct SurfacesData : BaseGlobalStruct
     Array1D<bool> SurfWinSlatAngThisTSDegEMSon;      // flag that indicate EMS system is actuating SlatAngThisTSDeg
     Array1D<Real64> SurfWinSlatAngThisTSDegEMSValue; // value that EMS sets for slat angle in degrees
     Array1D<bool> SurfWinSlatsBlockBeam;             // True if blind slats block incident beam solar
+    Array1D<int> SurfWinSlatsAngIndex;
```
Comment (mjwitte) [inline src/EnergyPlus/DataSurfaces.hh:1541]: 6 new `SurfWin*` arrays to avoid recomputing values.

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 Nice work here. I have a little more testing to do, but looks good so far.

**Review by mjwitte — APPROVED** [review]
@xuanluo113 Thanks for addressing the comments. Since the performance tests don't have any blinds, I dug up the notorious file from last year with 100s of windows with blinds in a large zone. It's still warming up after running for 12 hrs. So, instead I took PurchAirWindowBlind_BlockBeamSolar and ran that for 10yrs and 20yrs with develop and this branch. This branch runs about 10% faster. Nice! Merging.

**Pull request merged**